text/0000-dynamic-attributes.md

```diff
@@ -0,0 +1,150 @@
+
+## Example
+```js
+// x/myComponent.js
+export default class MyComponent extends LightningElement {
+    childProps = { 
+        name: 'My Child',
+    };
+}
+```
+```html
+// x/myComponent.html
+<template>
+    <x-child lwc:spread={childProps}></x-child>
+</template>
+```
+
+The new `lwc:spread` directive passes `childProps` as properties to `x-child`.
+
+## Motivation
+
+In LWC, there is no way to pass props to a child component when the component author
+doesn't know in advance all the props that are needed.
+
+One such usecase is `lwc:dynamic` which allows component authors to lazily instantiate components. However,
+the props that we pass to the component can't be dynamic: we need to pass them in the template HTML itself.
+
+This is a problem when the component constructor is changed at runtime: the new constructor
+might need a different set of properties than the earlier one.
+
+We can, of course, not pass any props and once the element renders, query the
+element and set the props programmatically. We lose reactivity this way and also force
+the component to handle `null` props which might not be desirable.
+
+This proposal allows not just `lwc:dynamic` but all other elements to accept an object which is bound as props
+at runtime.
+
+### Prior Art
+- React ([JSX Spread Operator](https://reactjs.org/docs/jsx-in-depth.html#spread-attributes))
+- Vue ([`v-bind`](https://v3.vuejs.org/api/directives.html#v-bind))
+- Svelte ([Spread Props](https://svelte.dev/tutorial/spread-props))
+- Lit ([Discussion for support](https://github.com/lit/lit/issues/923), [PR](https://github.com/lit/lit/pull/1960))
+
+## Detailed Design
+
+A new LWC directive, `lwc:spread` will be introduced that can be used to spread the props to children elements.
+
+### Structure
+It accepts an object with keys as the property names and values.
+
+This object is merged with props passed in the template (essentially `Object.assign`).
+
+There is one difference however: the set of properties passed in the template doesn't change but it changes in `lwc:spread` i.e.
+a property can't be present in one render and absent in the next render if it is defined in the template. With
+`lwc:spread` though, it can be present in one render and not be present in the next. See example below.
+
+**Example**
+
+```html
+<template>
+    <x-child name={name}></x-child> <!-- in first render -->
+    <x-child></x-child> <!-- in second render, name is missing. This is not possible with template compiler -->
+</template>
+```
+
+```html
+<template>
+    <x-child lwc:spread={props}>
+</template>
+```
+
+```js
+export default class App extends LightningElement {
+    handleClick() {
+        this.props = showName ? {name: 'lwc'} : {}; // in one render we're passing `name` prop, but in another one there's no `name` property.
+    }
+
+}
+```
+
+In cases where a property from earlier render is missing, we have two options:
+1. Explicity set it to `undefined` or `null`.
```

Comment:
setting a prop to undefined might be problematic... some properties will try to stringify that, and will result on `"undefined"` as the value. I think we should stay away from prescribing anything here, just saying that whatever it is in the object, will be set into the element, in a one way street, no diffing, no segregation of values, etc. just straight Object.assign().

I also think that having the desugaring syntax would help a lot.

Reply:
The problem with this approach is it's dependent on the previous render as well, not just what is passed to `lwc:spread`.

E.g.
```html
<x-child lwc:spread={props}></x-child>
<button onclick={handleClick1}>Button 1</button>
<button onclick={handleClick2}>Button 2</button>
<button onclick={handleClick3}>Button 3</button>
<button onclick={handleClick4}>Button 4</button>
```
```js
handleClick1() {  this.props = {name: "alice"}; }
handleClick2() {  this.props = {age: 10};}
handleClick3() {  this.props = {name: "bob"};}
handleClick4() {  this.props = {age: 37};}
```
In the above component, depending on the _order_ in which the buttons are clicked, we get different behavior.